### ISSUE TITLE
Wrong initialization of state from measurements

### ISSUE THREAD

**Issue by mithi** [ColinShaw/cpp-unscented-kalman-filters#1]
Hello @ColinShaw 

I just noticed that in your file 
https://github.com/ColinShaw/cpp-unscented-kalman-filters/blob/master/src/ukf.cpp

in your function `void UKF::InitializeMeasurement()`
in lines `119 - 123` 
you initialized the state as follows:
```
    x << rho * cos(phi),
         rho * sin(phi),
         rho_dot, 
         phi,
         0.0;
```
From what I've gathered, `phi` is not the same as `yaw` and `rho_dot` is not the same as `v`

https://discussions.udacity.com/t/mapping-from-radar-measurement-to-state-px-py-v-yaw-yaw-rate/237746

**Comment by ColinShaw** [comment]
Hello Mithi, thank you for pointing this out!  Interesting problem, isn't it.  Seemed to me at the time, particularly since it is just an initial state, that rho dot would be a reasonable estimator of v.  I suppose the idea being that the two vectors were reasonably close to the same direction, though I suppose this is a pretty big leap of faith to be true.  Similarly with the use of phi.  Now that you point it out, this seems inconsistent and flawed.  Seems like simply computing x and y from the radar measurement is probably best, and allow the rest to default to zero.

**Comment by ColinShaw** [comment]
Closing the issue as I updated it.

**Comment by mithi** [comment]
@ColinShaw Oh! I see you're right it can be a reasonable estimator! I didn't think of it that way!
As for me, I'm still stuck on my own code right now! Currently, my P is exploding. (NaN, inF numerical instability), even though I think I'm initializing all my variables and normalizing all my angles. Hope I find the the solution soon!

**Comment by ColinShaw** [comment]
I had some of that happen as well, particularly with the second data set.  I put the time threshold check in my UpdateTime() method to help with that.  It introduces minor error but gets rid of the catastophic error.   I had other issues with stability earlier, and while I had read a few things people tried with alternatives to the way the matrix square root is calculated, that wasn't the problem I experienced.  It isn't all that hard to choose values for your noise and initialization values that cause stability problems.  Since those are hard to guess, I used the data files to calculate estimates based on the residuals of the noisy data and ground truth data.
